docs: precision issues around price; short and long term solution (#5552)
osmosis-labs/osmosis: x/concentrated-liquidity/README.md

a precision of 18 and a very low price level relative to
the quote asset that has precision of 6 (e.g `uosmo` or `uusdc`).

The true price of PEPE in USDC terms is `0.0000009749`.

In the "on-chain representation", this would be:
`0.0000009749 * 10**6 / 10**18 = 9.749e-19`

Note that this is below the minimum precision of `sdk.Dec`.

Additionally, there is a problem with tick to sqrt price conversions
where at small price levels, two sqrt prices can map to the same
tick.

As a workaround, we have decided to limit min spot price to 10^-12
and min tick to `-108000000`. It has been shown at at price levels
below 10^-12, this issue is most apparent. See this issue for details:
<https://github.com/osmosis-labs/osmosis/issues/5550>

Now, we have a problem that we cannot handle pairs where
the quote asset has a precision of 6 and the base asset has a
precision of 18.

Note that this is not a problem for pairs where the quote asset
has a precision of 18 and the base asset has a precision of 6.
E.g. OSMO/DAI.

### Solution

At launch, pool creation is permissioned. Therefore, we can
ensure correctness for the initial set of pools.

> Long term, we will implement a wrapper contract around concentrated liquidity
> that will handle the precision issues and scale the prices to be all in the
> desired precision of 6.

Reviewer: ```suggestion
Long term, we will implement a wrapper contract around concentrated liquidity
that will handle the precision issues and scale the prices to all have a precision of at most 12.
```

all incentive records for a given pool ID and min uptime index by performing prefix iteration.

## Precision Issues With Price

There are precision issues that we must be considerate of in our design.

Consider the balancer pool between `uarb` and `uosmo`:

```bash
osmosisd q gamm pool 1011
pool:
  '@type': /osmosis.gamm.v1beta1.Pool
  address: osmo1pv6ffw8whyle2nyxhh8re44k4mu4smqd7fd66cu2y8gftw3473csxft8y5
  future_pool_governor: 24h
  id: "1011"
  pool_assets:
  - token:
      amount: "101170077995723619690981"
      denom: ibc/10E5E5B06D78FFBB61FD9F89209DEE5FD4446ED0550CBB8E3747DA79E10D9DC6
    weight: "536870912000000"
  - token:
      amount: "218023341414"
      denom: uosmo
    weight: "536870912000000"
  pool_params:
    exit_fee: "0.000000000000000000"
    smooth_weight_change_params: null
    swap_fee: "0.002000000000000000"
  total_shares:
    amount: "18282469846754434906194"
    denom: gamm/pool/1011
  total_weight: "1073741824000000"
```

Let's say we want to migrate this into a CL pool where `uosmo` is the quote
asset and `uarb` is the base asset.

Note that quote asset is denom1 and base asset is denom0.

> We want quote asset to be `uosmo` so that limit oredrs on ticks

Reviewer 2: ```suggestion
We want quote asset to be `uosmo` so that limit orders on ticks
```

> We want quote asset to be `uosmo` so that limit oredrs on ticks

Reviewer 3: ```suggestion
We want quote asset to be `uosmo` so that limit orders on ticks
```

have tick spacing in terms of `uosmo` as the quote.

Note:
- OSMO has precision of 6. 1 OSMO = 10**6 `uosmo`

> - ARB has precision of 18. 1 ARB = 10**18 `uarb`

Reviewer 3: Would this technically be `uarb`? My understanding was that the `u` represented µ/micro/10^-6 (might be mistaken on this lol)

Author: My interpretation of `u` was the base AMM unit. However, I'm not sure so I will change this to something else

Reviewer 4: Alpin is correct in it referring to micro

Author: addressed

Therefore, the true price of the pool is:
```python
>>> (218023341414 / 10**6)  / (101170077995723619690981 / 10**18)
2.1550180224553714
```

However, in our core logic it is represented as:

```python
218023341414 / 101170077995723619690981
2.1550180224553714e-12
```

or

```python
osmosisd q gamm spot-price 1011 uosmo ibc/10E5E5B06D78FFBB61FD9F89209DEE5FD4446ED0550CBB8E3747DA79E10D9DC6

> spot_price: "0.000000000002155018"

Reviewer 2: just curious; so we do `0.000000000002155018 * (10**12)` in client side to display the spot price?

```

As a protocol, we need to accomodate prices that are very far apart.
In the example above, the difference between `10**6 and 10**18`

Most of the native precision is 10**6. However, most of the ETH
precision is 10**18.

This starts to matter for assets such as `upepe`. That have
a precision of 18 and a very low price level relative to
the quote asset that has precision of 6 (e.g `uosmo` or `uusdc`).

The true price of PEPE in USDC terms is `0.0000009749`.

> In the "on-chain representaiton", this would be:

Reviewer 2: ```suggestion
In the "on-chain representation", this would be:
```

`0.0000009749 * 10**6 / 10**18 = 9.749e-19`

> Noe that this is below the minimum precision of `sdk.Dec`.

Reviewer 2: ```suggestion
Note that this is below the minimum precision of `sdk.Dec`.
```

Reviewer 2: So the min and max precision of sdk.Dec is 10^-18 and 10^18 ?

Author: Too be clear, min value is 10**(-18) with a defined precision of 18.

The max value is some other number defined by this:
https://github.com/osmosis-labs/cosmos-sdk/blob/1cfa6927834d2d3d392aeaf59fe929a057984e92/types/decimal.go#L29

Author: I will reword this. I think this isn't worded precisely

Additionally, there is a problem with tick to sqrt price conversions
where at small price levels, two sqrt prices can map to the same
tick.

As a workaround, we have decided to limit min spot price to 10^-12
and min tick to `-108000000`. It has been shown at at price levels
below 10^-12, this issue is most apparent. See this issue for details:
<https://github.com/osmosis-labs/osmosis/issues/5550>

Now, we have a problem that we cannot handle pairs where
the quote asset has a precision of 6 and the base asset has a
precision of 18.

Note that this is not a problem for pairs where the quote asset
has a precision of 18 and the base asset has a precision of 6.

> E.g. OSMO/DAI.

Reviewer 2: i'm trying to understand this a little more

so in the above example, 
```
osmo quote (6th precision) & arb base (18th prevision)   
spot price = (218023341414 / 10**6) / (101170077995723619690981 / 10**18) 
spot price = 2.1550180224553714
```

Why would this not work?

Author: Because in AMM units, this is:
```
>>> 218023341414 / 101170077995723619690981
2.1550180224553714e-12
```

It does not account for precision. As a result, we are right at the boundary of the spot price that can be represented without issues

Reviewer 2: i see, new min limit is 10^-12. 

but why don't we do (quote / precision) / (base / precision) on instead of just quote / base, on chain side?

Author: This is how precisions for these assets were historically defined. So we must normalize them to be in the same domain
